ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/RaceIterator.hs

```diff
@@ -0,0 +1,151 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE PatternSynonyms     #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Test.Ouroboros.Consensus.ChainGenerator.RaceIterator (
+    Race (Race, UnsafeRace)
+  , RaceLbl
+  , init
+  , initConservative
+  , next
+  , nextConservative
+  ) where
+
+import           Control.Monad (when)
+import           Data.Proxy (Proxy (Proxy))
+import           Prelude hiding (init)
+import qualified Test.Ouroboros.Consensus.ChainGenerator.BitVector as BV
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Counting as C
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Delta (Delta),
+                     Kcp (Kcp), Scg (Scg))
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Slot as S
+import           Test.Ouroboros.Consensus.ChainGenerator.Slot (E (SlotE), S)
+
+-----
+
+-- | See 'Race'
+data RaceLbl
+
+-- | A window whose last slot contains the @k+1@st active slot in it
+newtype Race base = UnsafeRace (C.SomeWindow RaceLbl base SlotE)
+  deriving (Eq, Read, Show)
+
+pattern Race :: C.SomeWindow RaceLbl base SlotE -> Race base
+pattern Race x <- UnsafeRace x
+
+{-# COMPLETE Race #-}
+
+-----
+
+-- | Yields the race window starting at position 0 of the given
+-- vector if the @k+1@ active slot exists.
+init :: Kcp -> C.Vector base SlotE S -> Maybe (Race base)
+init (Kcp k) v = do
+    -- find the @k+1@st active slot in the given race window
+    case BV.findIthEmptyInV S.inverted v (C.Count k) of
+        BV.NothingFound       -> Nothing
+        BV.JustFound kPlus1st ->
+            Just
+         $! UnsafeRace
+         $  C.withWindowBetween
+                (C.lengthV v)
+                (C.Lbl @RaceLbl)
+                (C.Count 0)
+                kPlus1st
+
+-- | @initConservative@ is like @init@, but assumes that any Race window
+-- spilling the vector length has all the excedent slots active.
+--
+-- This function is only safe to call when we know that 'init' would return
+-- @Nothing@.
+--
+initConservative :: Scg -> Delta -> C.Contains SlotE outer inner -> Race inner
+initConservative (Scg s) (Delta d) win =
+    UnsafeRace
+  $ C.withWindow
+        (C.windowSize win)
+        (C.Lbl @RaceLbl)
+        (C.Count 0)
+        (C.Count (s - d))
```

Comment:
The size of the window needs to extend past the container size in order for the conservative Race window to have `k+1` active slots. Moreover, it should be counted how many active slots there are in the overlapping portion of the window, otherwise the race window might be too small.

Reply:
This code has been rewritten since this comment was made and it no longer applies.